### PR TITLE
Mesh 1039/Allow compliance management based on sender/receiver's did

### PR DESCRIPTION
**Pull request by maxsam4**
1) Remove special cases for issuance/redemption on the basis of the treasury.
2) Add a new rule type to specify the sender/receiver identity (specific static or dynamic treasury). These rules are meant to be used to specify issuance/redemption rules. CM can now also be used as allowlist/blocklist TM.
3) Issuance function now only issues to the treasury (fallback to the owner). Compliance rules are not checked during issuance.

### CODE REVIEW

File: primitives/src/predicate.rs [hunk 1]
```diff
@@ -108,6 +130,26 @@ pub fn run(rule: &Rule, context: &Context) -> bool {
...
+        }
+    }
+}
+
+// MatchPredicate
+// ======================================================
+
+/// It matches `id` with treasury in the context.
+#[derive(Clone, Debug)]
+pub struct MatchPredicate<'a> {
```
Comment (vkomenda) [inline primitives/src/predicate.rs:144]: `Match` sounds a bit generic. It it makes a statement about the treasury identity, maybe replace `Match` with `Treasury` or `TreasuryIdentity`?

File: pallets/compliance-manager/src/lib.rs [hunk 2]
```diff
@@ -824,38 +835,33 @@ impl<T: Trait> ComplianceManagerTrait<T::Balance> for Module<T> {
...
+        if asset_rules.is_paused {
             return Ok(ERC1400_TRANSFER_SUCCESS);
         }
         for active_rule in asset_rules.rules {
-            let mut rule_satisfied = false;
             if let Some(from_did) = from_did_opt {
-                rule_satisfied = treasury_sender
-                    || Self::are_all_rules_satisfied(ticker, from_did, &active_rule.sender_rules);
-                if !rule_satisfied {
+                if !Self::are_all_rules_satisfied(
```
Comment (satyamakgec) [inline pallets/compliance-manager/src/lib.rs:845]: Does that mean the treasury_did should also possess all the valid claims as to any sender?  &  issuance will only work if the receiver has the `TargetIdentity::specific(identity)`  claim.

Comment (maxsam4) [reply pallets/compliance-manager/src/lib.rs:845]: `TargetIdentity::specific(identity)` is not a claim, it's a rule. The receiver needs no claims. However, there should be separate compliance manager rule that says something along the lines of "Bypass all receiver restrictions if the sender is treasury".

Comment (satyamakgec) [reply pallets/compliance-manager/src/lib.rs:845]: I think it should be something like "Bypass all the restrictions when the sender is treasury or receiver is treasury". We still need to impose the restrictions on the receiver if the sender is the treasury or vice-versa. 

Even my main concern with the current code is that Treasury did have to pass all the rules applied to the sender itself and that defeat the purpose of giving the special case for the treasury.

Comment (maxsam4) [reply pallets/compliance-manager/src/lib.rs:845]: > I think it should be something like "Bypass all the restrictions when the sender is treasury or receiver is treasury". We still need to impose the restrictions on the receiver if the sender is the treasury or vice-versa.

We don't need to impose anything. The issuer can create rules as per their regulatory requirements. FWIW, I meant the rule to be "Sender is treasury". This implies "Bypass all receiver restrictions if the sender is treasury". To cover the redemption case, the issuer can add another rule saying "Receiver is treasury" (Implicit - no rules for sender).

> Even my main concern with the current code is that Treasury did have to pass all the rules applied to the sender itself and that defeat the purpose of giving the special case for the treasury.

This change removes the special case of the treasury. The issuer must add proper compliance rules before issuing. The dApp can batch both calls together for the issuer.

Comment (satyamakgec) [reply pallets/compliance-manager/src/lib.rs:845]: I still don't understand why we need the special rule type? I can still issue asset to the treasury without having the `TargetIdentity:Treasury` rule in place.

Comment (maxsam4) [reply pallets/compliance-manager/src/lib.rs:845]: This rule is not to cover issuance. It is to facilitate distribution to investors after the tokens are already issued to the treasury.

This PR separates the process of issuance from distribution. Issuance is always done to the treasury and then those tokens are distributed to the investors.

File: pallets/asset/src/lib.rs [hunk 3]
```diff
@@ -707,129 +706,21 @@ decl_module! {
...
-        /// * `ticker` Ticker of the token.
-        /// * `investor_dids` Array of the DID of the token holders to whom new tokens get issued.
-        /// * `values` Array of the Amount of tokens that get issued.
-        ///
-        /// # Weight
-        /// `300_000 + 400_000 * issue_asset_items.len().max(values.len())`
-        #[weight =
-            300_000 + 400_000 * u64::try_from(issue_asset_items.len()).unwrap_or_default()
-        ]
-        pub fn batch_issue(origin, issue_asset_items: Vec<IssueAssetItem<T::Balance>>, ticker: Ticker) -> DispatchResult {
```
Comment (adamdossa) [inline pallets/asset/src/lib.rs:741]: Is the idea to use utility batch functions instead?

**Pull request merged**